keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2248 @@
+
+#### Access control
+
+This future step adds support for controlling the access to available QoS-class
+resources.
+
+Access control could be achieved e.g. by extending ResourceQuotaSpec which
+would implement restrictions based on the namespace.
+
+```diff
+ // ResourceQuotaSpec defines the desired hard limits to enforce for Quota.
+ type ResourceQuotaSpec struct {
+     // hard is the set of desired hard limits for each named resource.
+     Hard ResourceList
+     // A collection of filters that must match each object tracked by a quota.
+     // If not specified, the quota matches all objects.
+     Scopes []ResourceQuotaScope 
+     // scopeSelector is also a collection of filters like scopes that must match each
+     // object tracked by a quota but expressed using ScopeSelectorOperator in combination
+     // with possible values.
+     ScopeSelector *ScopeSelector
++   // QOSResources contains the desired set of allowed QoS resources.
++   // +optional
++   QOSResources QOSResourceQuota
+ }
+
++// QOSResourceQuota contains the allowed QoS resources.
++type QOSResourceQuota struct {
++       // Pod contains the allowed QoS resources for pods.
++       // +optional
++       Pod []AllowedQOSResource
++       // Container contains the allowed QoS resources for pods.
++       // +optional
++       Container []AllowedQOSResource
++}
+
++// AllowedQOSResource specifies access to one QoS resources type.
++type AllowedQOSResource struct {
++       // Name of the resource.
++       Name QOSResourceName
++       // Allowed classes.
++       Classes []AllowedQOSResourceClass
++}
+
++// AllowedQOSResourceClass specifies one allowed class of a QoS resource and
++// possible limits for its usage.
++type AllowedQOSResourceClass struct {
++       // Name of the class.
++       Name string
++       // Capacity is the hard limit for usage of the class.
++       // +optional
++       Capacity int64
++}
+
+ // ResourceQuotaStatus defines the enforced hard limits and observed use.
+ type ResourceQuotaStatus struct {
+        ...
+        // Used is the current observed total usage of the resource in the namespace
+        // +optional
+        Used ResourceList
++       // QOSResources contains the enforced set of available QoS resources.
++       // +optional
++       QOSResources QOSResourceQuota
++       // QOSResourcesUsage contains the observed usage of QoS resources in the namespace.
++       // +optional
++       QOSResourcesUsage QOSResourceQuota
+ }
+```
+
+#### Scheduler improvements
+
+The first implementation phase only adds basic filtering of nodes based on
+QoS-class resources and node scoring is not altered. However, the relevant
+scheduler plugins (e.g. NodeResourcesFit and NodeResourcesBalancedAllocation)
+could be extended to do scoring based on the capacity, availability and usage
+of QoS-class resources on the nodes.
+
+#### Kubelet-initiated pod eviction
+
+QoS-class resources available on a node are dynamic in the sense that they may
```

Comment:
if the ultimate intention of qos classes are to be dynamic, i think it makes sense to impelement it the right way the first time.  I think cpu/memory being dynamic is a bit different, because these fields are still owned totally by the kubelet, not by the cri impl. Thus, discovering and handling changes will work differently for each

Reply:
Yes, I agree with @haircommander. I don't see much reason for forcing them to be static. And since they are managed by the runtime you cannot really prevent them from being dynamic, in practice, imo